Add HTTP webserver interface

15532th opened the pull request:
This PR implements enhancements discussed in #1, namely serving token as a JSON endpoint. It is meant to be a starting point for discussing the HTTP interface, specifically endpoints names and responses formats, so all the new code is placed into a separate file to avoid affecting any existing functionality.

Token generation runs independently from the webserver process, launching an update every 12 hours by default. Webserver uses port `8080` (configurable), with the latest token value being available on the `/token` endpoint:

```bash
$ curl 127.0.0.1:8080/token
{"updated": 1724445426, "potoken": "...", "visitor_data": "..."}
```

Additionally, it is possible to initiate immediate unscheduled token generation by accessing `/update` endpoint. While handy for testing, this feature seems to add more complications than it's worth, so it probably should be removed, unless it finds some practical use.

unixfox commented:
thank you for this @15532th. I think it would be greatly appreciated if you could integrate it inside the main script. this way we can keep the logic common between the server web API and the one shot script.

15532th committed:
Replace use of TemporaryDirectory with mkdtemp

Turns out TemporaryDirectory initiates cleanup on interpreter exit even when it is used without context manager, which might interfere with nodriver trying to delete the directory at the same time.

15532th committed:
Handle possibility of "#movie_player" selector timing out

It might happen if loaded page does not contain a video player or fails to load at all.

Explicitly specify 10 seconds timeout for the selector, wrap it in a try-catch block, add separate error messages for the selector and handler timeouts.

15532th committed:
Add named logger in potoken.py, update usages

It allows client code to control logging format and level when the module is used as a library and not a standalone script.

15532th committed:
Abort token update if it takes longer than ten minutes

If connection to the launched browser gets interrupted or the browser itself exits abruptly before the video tab got loaded, the update procedure might hang up indefinitely.

15532th committed:
Update index.py to use potoken.py code

15532th committed:
Rewrite token generation, add web-server

15532th commented:
Updated the main script. Output on successful token retrieval should be the same, however it might print different messages on errors.

I also fixed a few issues I came across while updating it, see individual commit messages for comments on the changes.

unixfox commented:
@WardPearce Could you help me review? Thank you :)

WardPearce commented:
> @WardPearce Could you help me review? Thank you :)

more then happy to, I'll have a look over tmrw

WardPearce commented:
@unixfox @15532th 
Is their any reason why we couldn't use a micro web framework like litestar, starlette, fastapi etc? Would be more readable & provide additional functionality like automatic API documentation etc etc. (wsgiref itself doesn't appear to be considered production ready too & is a wrapper of https://docs.python.org/3/library/http.server.html)

Wouldn't updating Po tokens be better done using a in memory cache, what will expire after X seconds. Meaning that after every hour or so the request is slightly slower or possibly using aiocron (https://github.com/gawel/aiocron) if we must schedule in the background.

15532th commented:
@WardPearce
I do have a couple of reasons I don't want to do this. Firstly, for an endpoint that is going to be accessed by a single client sitting on the same machine, making at worst one request per hour, it feels like an overkill to use production-ready frameworks capable of handling hundreds of RPS to achieve the same result as a few dozens of lines of pure Python code. Same goes for using in-memory cache to hold a couple of short strings (for all the added resource overhead and complexity it would only make things slower, I believe) and dealing with a cron-like interface to achieve something that can be done by a single coroutine with a while loop.

The second reason is that the typical environment I expect myself to end up using this token-retrieving functionality in, which is the end user machine, often running Windows. I don't want to ask my users to deal with complicated dependencies installation or setting up and running a docker image. In the current state of the token generator it only has a single pure Python dependency, which makes it easy to publish on pypi or package as an executable without running into obscure cross-platform compatibility issues (I am leaving out the biggest dependency, the Chrome itself, because a user machine is quite likely going to have a fresh version installed).

Regarding the "run update inside the Nth requests" vs "do updates in background" choice, my reasons for deciding on the latter were the following: while the usual delay on a fast machine would be only a few seconds, the worst case is currently a 30 seconds timeout, which some clients might not be prepared to handle, since it is not a delay you would normally expect from a web-service running on localhost. And if the issue that caused the timeout was only temporary, it means client has received an unnecessarily error while it could have been operating fine if it was served the old cached version of the token until a new update succeeds.

unixfox commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+    token regeneration by accessing http://127.0.0.1:8080/update
+    '''
+    parser = argparse.ArgumentParser(description=description, formatter_class=argparse.RawDescriptionHelpFormatter)
+    parser.add_argument('--update-interval', '-u', type=int, default=3600 * 12,
```
could you make it so it refreshes the token every hour by default? this is much better for privacy.

15532th replied:
Personally I would expect the backend to be smart enough to merge history from multiple tokens, considering neither IP nor hardware fingerprint is changing between updates. However a hour does seems to be a more reasonable number then twelve, considering no retry on failure is implemented.

unixfox commented:
The point in having an API is to fetch the po_token from an external tool that is not designed to do it by itself.

A po_token currently works for an indefinite amount of time, it's not a token that will expire after 1 hour. So serving a po_token generated from one hour ago is perfectly fine.

I do share the same vision as @15532th, this API won't be used at high scale and won't be accessible from the whole internet. Worst case scenario, someone runs 10 Invidious processes, it's just 10 requests done per hour, it's meaningless.

15532th committed:
Set default token update interval to 1 hour

unixfox commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+    def get_potoken(self) -> Tuple[str, list, str]:
+        token = self._potoken_extractor.get()
+        if token is None:
+            status = '502 Bad Gateway'
```
`503 Service Unavailable` makes more sense here: https://stackoverflow.com/a/12281287

WardPearce commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+    daemon_threads = True
+
+
+class PotokenServer:
```
Server logic should probably be separated from  extractor logic. Possible structure
```
- session_generator/
   - __init__ - Contains CLI entry point & server entry point
   - server
   - extractor
```

15532th replied:
I would like to postpone packaging this as a module until it comes to publishing it to pypi.
It's quite handy to have it as a single file that can be run from any directory without installing, so I would rather keep it in this state while possible.

unixfox replied:
https://github.com/iv-org/youtube-trusted-session-generator/pull/6#discussion_r1765773287

WardPearce commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+        self.update_interval: float = update_interfal
+        self.profile_path = mkdtemp()  # cleaned up on exit by nodriver
+        self._loop = loop
+        self._token_info: Optional[dict] = None
```
Use dataclass to define token info

WardPearce commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+            logger.info('update was succeessful')
+            return True
+
+    async def _send_handler(self, event: nodriver.cdp.network.RequestWillBeSent):
```
Return notation, not exactly required but some parts of the code has None notated and others don't, prob worth notating the same way consistently

WardPearce commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+        try:
+            player = await tab.select("#movie_player", 10)
+        except asyncio.TimeoutError:
+            logger.warning(f'update failed: unable to locate video player on the page')
```
No need to be a fstring

WardPearce commented on potoken.py:
```diff
@@ -0,0 +1,255 @@
+
+    def request_update(self) -> Tuple[str, list, str]:
+        status = '200 OK'
+        headers = [("Content-Type", "text/plain")]
```
Consistently use `"` or `'` unless for escaping reasons.

15532th committed:
Change cold start error status code for the /token endpoint to 503

15532th committed:
Fix linter warnings

Remove unnecessarily f-strings and double quotes, fix a few other
issues.

15532th committed:
Add missing typehints

15532th committed:
Use dataclass to represent token_info

unixfox commented on potoken.py:
Could you rename the file to something like webapi.py?
And change the index.py to potoken.py

Make sure to change the entrypoint in the Dockerfile and README.

This way the filenames will make more sense to the user.

15532th replied:
Perhaps it would be better to move index.py functionality into the potoken.py, making it a single file script?

Single file script has a nice property of being able able to run with any current working directory, and if we get to renaming files, it would make sense to follow @WardPearce suggestion to separate web-server and extractor code, and make whole thing a proper package while at it.

unixfox replied:
I'm more in favor of doing like we did on https://github.com/iv-org/smart-ipv6-rotator

Multiple separate files instead of a single file. Since mainly everyone is going to use the script in Docker anyway.

WardPearce replied:
yea I agree, plus worse case someone needs to git clone the repo and the file structure will be the same.

15532th replied:
I updated the file structure combining both suggestions, let me know if you prefer different name for the package.

Two latest commits fix a couple of semi-related potential problems, they probably require a separate review.

WardPearce replied:
Yea I think its a lot more readable, personally I like the server cli & oneshot cli being combined so their is only one entry for the script.

15532th committed:
Duplicate index.py functionality in potoken.py

15532th committed:
Split potoken.py into multiple files

15532th committed:
Run first extraction before starting webserver

It leaves less time for client to hit 503 error on startup, but more importantly, it deals with race condition in startup sequence: if extractor_task gets interrupted after webserver startup was initiated but before make_server() returned, then potoken_server.stop() becomes a no-op, leaving background task with webserver to run indefinitely.

15532th committed:
Add a command line argument allowing to specify the path to a Chromium binary

unixfox committed:
bind webserver localhost + specific Dockerfile webserver + reformat doc
